docs/architecture/adr-063-core-module-api.md

```diff
@@ -0,0 +1,546 @@
+used to register both `Msg` and query services.
+
+#### Genesis
+
+The genesis `Handler` functions - `DefaultGenesis`, `ValidateGenesis`, `InitGenesis` and `ExportGenesis` - are specified
+against the `GenesisSource` and `GenesisTarget` interfaces which will abstract over genesis sources which may be single
+JSON files or collections of JSON objects that can be efficiently streamed. They will also include helper functions for
+interacting with genesis data represented by single `proto.Message` types.
+
+```go
+type GenesisSource interface {
+    ReadMessage(proto.Message) error
+    OpenReader(field string) (io.ReadCloser, error)
+    ReadRawJSON() (json.RawMessage, error)
+}
+
+type GenesisTarget interface {
+    WriteMessage(proto.Message) error
+    OpenWriter(field string) (io.WriteCloser, error)
+    WriteRawJSON(json.RawMessage) error
+}
+```
+
+All genesis objects for a given module are expected to conform to the semantics of a JSON object. That object may
+represent a single genesis state `proto.Message` which is the case for most current modules, or it may represent a set
+of nested JSON structures which may optionally be streamed as arrays from individual files which is the way the
+[ORM](./adr-055-orm.md) is designed. This streaming genesis support may be beneficial for chains with extremely large
+amounts of state that can't fit into memory.
+
+Modules which use a single `proto.Message` can use the `GenesisSource.ReadMessage` and `GenesisTarget.ReadMessage`
+methods without needing to deal with raw JSON or codecs.
+
+Modules which use streaming genesis can use the
+`GenesisSource.OpenReader` and `GenesisTarget.OpenWriter` methods. In the case of modules using the ORM, all JSON
+handling is done automatically by the ORM and a single line will be sufficient for registering all of that functionality
+(ex. `ormDb.RegisterGenesis(handler)`).
+
+Low-level `ReadRawJSON` and `WriteRawJSON` are provided in case modules use some other sort of legacy genesis logic
+we haven't anticipated.
+
+#### Begin and End Blockers
+
+The `BeginBlock` and `EndBlock` methods will simply take a `context.Context`, because:
+* most modules don't need Tendermint information other than `BlockInfo` so we can eliminate dependencies on specific
+Tendermint versions
+* for the few modules that need Tendermint block headers and/or return validator updates, specific versions of the
+runtime module will provide specific functionality for interacting with the specific version(s) of Tendermint
+supported
+
+In order for `BeginBlock`, `EndBlock` and `InitGenesis` to send back validator updates and retrieve full Tendermint
+block headers, the runtime module for a specific version of Tendermint could provide services like this:
+```go
+type ValidatorUpdateService interface {
+    SetValidatorUpdates(context.Context, []abci.ValidatorUpdate)
+}
+
+type BeginBlockService interface {
+    GetBeginBlockRequest(context.Context) abci.RequestBeginBlock 
+}
+```
```

Comment:
I don't think we should allow modules to deal with anything tendermint related. Tendermint is a single consensus protocol that the module shouldn't care about. Ideally sdk defines an interface or concrete type that base app/abci turns into a tendermint type. This way modules don't need to depend on tendermint when they are spun out.

Reply:
Is it fine if x/staking depends on tendermint? What I'm proposing here is that only x/staking would need that dependency and other modules wouldn't. We could also define some other validator update type in core or runtime too.

Reply:
Yes, this is fine. In the future when we decide to own the abci types ourselves and have a transition Layer for tendermint we can modify it. For now this isnt a worry nor should it block anything